Libraries/LibJS/Runtime/JSONObject.cpp

```diff
@@ -495,4 +503,60 @@ ThrowCompletionOr<Value> JSONObject::internalize_json_property(VM& vm, Object* h
     return TRY(call(vm, reviver, holder, PrimitiveString::create(vm, name.to_string()), value));
 }
 
+// https://tc39.es/proposal-json-parse-with-source/#sec-json.rawjson
+JS_DEFINE_NATIVE_FUNCTION(JSONObject::raw_json)
+{
+    auto& realm = *vm.current_realm();
+
+    // 1. Let jsonString be ? ToString(text).
+    auto json_string = TRY(vm.argument(0).to_string(vm));
+
+    // 2. Throw a SyntaxError exception if jsonString is the empty String, or if either the
+    // first or last code unit of jsonString is any of 0x0009 (CHARACTER TABULATION),
+    // 0x000A (LINE FEED), 0x000D (CARRIAGE RETURN), or 0x0020 (SPACE).
+    auto bytes = json_string.bytes_as_string_view();
+    if (bytes.is_empty())
+        return vm.throw_completion<SyntaxError>(ErrorType::JsonMalformed);
+
+    auto first_char = bytes[0];
+    auto last_char = bytes[bytes.length() - 1];
+
+    if (first_char == 0x0009 || first_char == 0x000A || first_char == 0x000D || first_char == 0x0020 || last_char == 0x0009 || last_char == 0x000A || last_char == 0x000D || last_char == 0x0020)
+        return vm.throw_completion<SyntaxError>(ErrorType::JsonMalformed);
+
+    // 3. Parse StringToCodePoints(jsonString) as a JSON text as specified in ECMA-404.
+    // Throw a SyntaxError exception if it is not a valid JSON text as defined in that specification,
+    // or if its outermost value is an object or array as defined in that specification.
+    auto json = JsonValue::from_string(json_string);
+    if (json.is_error())
+        return vm.throw_completion<SyntaxError>(ErrorType::JsonMalformed);
+
+    if (json.value().is_object() || json.value().is_array())
+        return vm.throw_completion<SyntaxError>(ErrorType::JsonMalformed);
```

Comment:
I don't think `JsonMalformed` is the right error message for this one. The JSON is well-formed, but it's the wrong type. Let's add a new error type to signify that the parsed JSON value cannot be an object or array.